File: src/services/visitService.js
import { ApiError, request } from "./http";

// THE SWAP. Every function below kept its name, its arguments and its return
// value; only the insides changed, from filtering an array in memory to asking
// a Spring Boot API the same question. Nothing above this file moved: not
// Dashboard, not Visits, not Billing, not MyVisits, not VisitDetail, not
// PatientDetail, not CaregiverDetail, not CaregiverVisit.
//
// That is the whole argument for a service layer, and it is now a fact about
// this repository rather than a claim about it.
//
// The 300ms fake delay is gone. It existed to make the mock behave like a
// network, and there is a network now.

// GET /api/visits, with the filters as query parameters. An absent filter means
// no restriction, which is what the server does too.
export const getVisits = async ({ status, q, caregiverId, patientId } = {}) =>
    request("/visits", { params: { status, q, caregiverId, patientId } });

// GET /api/visits/counts. Still its own request for the same reason it always
// was: the chips count the whole collection while the list shows one slice.
export const getVisitCounts = async () => request("/visits/counts");

// GET /api/visits/{id}.
//
// A missing visit comes back as undefined, NOT as a thrown error, because that
// is the contract VisitDetail was written against: it renders a not-found page
// when the value is falsy and a retryable failure when the call throws. Those
// are two different answers and the 404 is the one that is not a failure.
export const getVisitById = async (id) => {
    try {
        return await request(`/visits/${id}`);
    } catch (error) {
        if (error instanceof ApiError && error.status === 404) return undefined;
        throw error;
    }
};

// Filtering by id, never by name. Kept as named verbs because the two doors
// differ in AUTHORIZATION rather than in query: an admin reads any caregiver's
// visits this way, while a caregiver reading their own becomes
// GET /api/visits/mine, where the server takes the id from the principal.
//
// The guard stays client side as well as server side. It is not the security
// boundary, it just fails immediately instead of spending a round trip to be
// told the same thing.
export const getVisitsByCaregiver = async (caregiverId) => {
    if (caregiverId == null || caregiverId === "") {
        throw new Error("A caregiver id is required");
    }

    return getVisits({ caregiverId });
};

export const getVisitsByPatient = async (patientId) => {
    if (patientId == null || patientId === "") {
        throw new Error("A patient id is required");
    }

    return getVisits({ patientId });
};

// The mutations are sub-resources named for the event that causes them. The
// transition guards and the four field evidence check now run in the Java
// service: they used to run here, which meant they were suggestions.
//
// Note what is NOT sent: no caregiver id and no timestamps. The server stamps
// the clock, because a time the caller could author is not evidence, and with
// real authentication the actor comes from the token rather than the body.
export const checkInVisit = async (id, location) =>
    request(`/visits/${id}/check-in`, { method: "POST", body: location ?? {} });

export const checkOutVisit = async (id, { assessment, signature }) =>
    request(`/visits/${id}/check-out`, { method: "POST", body: { assessment, signature } });

export const supplyEvidence = async (id, { assessment, signature }) =>
    request(`/visits/${id}/evidence`, { method: "POST", body: { assessment, signature } });

export const submitClaim = async (id) =>
    request(`/visits/${id}/claim`, { method: "POST" });

File: backend/src/main/java/com/vera/api/service/VisitService.java
package com.vera.api.service;

import com.vera.api.domain.CapturedLocation;
import com.vera.api.domain.Visit;
import com.vera.api.domain.VisitStatus;
import com.vera.api.repository.VisitRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * The domain service: where the rules live.
 *
 * THIS IS THE POINT OF THE PORT. The evidence check and every status transition
 * guard used to run in the browser, which meant they were suggestions: anything
 * that could reach the service could move a visit straight to "ready to bill"
 * without a signature. A rule the client enforces is not a rule. They run here
 * now, and the client keeps its copy only so the UI can grey out a button
 * before the round trip.
 *
 * Not in the controller, which only speaks HTTP, and not in the entity, which
 * only records what was decided. A rule split across layers is a rule nobody
 * can find.
 */
@Service
public class VisitService {

    /**
     * DEMO BEHAVIOUR, carried over from the mock deliberately. A real check-out
     * stamps Instant.now(); this stamps ninety minutes after the check-in so a
     * visit completed during a two minute demo still bills a plausible number of
     * hours. It is the one place this service tells a small, clearly labelled
     * lie, and it is labelled here rather than hidden behind a constant name.
     */
    private static final long DEMO_VISIT_MINUTES = 90;

    private final VisitRepository visits;

    public VisitService(VisitRepository visits) {
        this.visits = visits;
    }

    @Transactional(readOnly = true)
    public List<Visit> search(VisitStatus status, String q, Long caregiverId, Long patientId) {
        // The LIKE pattern is built here rather than in the query so the
        // repository stays a plain comparison. Blank or whitespace-only is not
        // a search: it must mean "everything", not "match the empty string".
        String namePattern = (q == null || q.isBlank())
                ? null
                : "%" + q.trim().toLowerCase() + "%";

        return visits.search(status, namePattern, caregiverId, patientId);
    }

    /** Throws rather than returning an Optional, so the 404 is decided once. */
    @Transactional(readOnly = true)
    public Visit getById(Long id) {
        return visits.findByIdWithRelations(id)
                .orElseThrow(() -> new NotFoundException("Visit " + id + " not found"));
    }

    /**
     * The counts behind the filter chips. Every status appears, including the
     * ones with no visits: a chip reading "billed (0)" is information, while a
     * missing key would make the client invent a zero it was never told.
     */
    @Transactional(readOnly = true)
    public VisitCounts counts() {
        Map<VisitStatus, Long> byStatus = new EnumMap<>(VisitStatus.class);
        for (VisitStatus status : VisitStatus.values()) {
            byStatus.put(status, 0L);
        }

        long total = 0;
        for (Object[] row : visits.countByStatus()) {
            VisitStatus status = (VisitStatus) row[0];
            long count = (Long) row[1];
            byStatus.put(status, count);
            total += count;
        }

        return new VisitCounts(total, byStatus);
    }

    /** total plus a count per status, which is what the chips render. */
    public record VisitCounts(long total, Map<VisitStatus, Long> byStatus) {
    }

    /**
     * The caller supplies the location because the device is the only authority
     * on where it is. The SERVER stamps the clock, because a time the caller
     * could author is not evidence.
     */
    @Transactional
    public Visit checkIn(Long id, CapturedLocation location) {
        Visit visit = getById(id);
        requireStatus(visit, VisitStatus.SCHEDULED, "check in");

        visit.recordCheckIn(Instant.now(), location);
        return visit;
    }

    @Transactional
    public Visit checkOut(Long id, String assessment, String signature) {
        Visit visit = getById(id);
        requireStatus(visit, VisitStatus.IN_PROGRESS, "check out");

        Instant checkOutTime = visit.getCheckInTime()
                .plusSeconds(DEMO_VISIT_MINUTES * 60);

        visit.recordCheckOut(checkOutTime, blankToNull(assessment), blankToNull(signature));
        visit.setStatus(gradeEvidence(visit));
        return visit;
    }

    /**
     * Supplying what was missing. Merges rather than replaces: sending nothing
     * for a field keeps what is already there, so this can only ever add
     * evidence and never erase it.
     */
    @Transactional
    public Visit supplyEvidence(Long id, String assessment, String signature) {
        Visit visit = getById(id);
        requireStatus(visit, VisitStatus.NEEDS_REVIEW, "supply evidence to");

        visit.recordEvidence(
                firstPresent(blankToNull(assessment), visit.getAssessment()),
                firstPresent(blankToNull(signature), visit.getSignature()));
        visit.setStatus(gradeEvidence(visit));
        return visit;
    }

    @Transactional
    public Visit submitClaim(Long id) {
        Visit visit = getById(id);
        requireStatus(visit, VisitStatus.READY_TO_BILL, "submit a claim for");

        visit.recordClaim("clm_mock_" + visit.getId(), Instant.now());
        return visit;
    }

    /**
     * THE EVIDENCE RULE. Four fields, all required, and no other way to reach
     * READY_TO_BILL. There is deliberately no method that sets it directly and
     * no admin override: a visit missing a signature is held until a signature
     * exists, because clicking "resolve" on a missing signature does not create
     * one.
     */
    private VisitStatus gradeEvidence(Visit visit) {
        boolean complete = visit.getCheckInTime() != null
                && visit.getCheckOutTime() != null
                && hasText(visit.getAssessment())
                && hasText(visit.getSignature());

        return complete ? VisitStatus.READY_TO_BILL : VisitStatus.NEEDS_REVIEW;
    }

    private static void requireStatus(Visit visit, VisitStatus required, String action) {
        if (visit.getStatus() != required) {
            throw new IllegalTransitionException(
                    "Cannot " + action + " a visit that is " + visit.getStatus().getLabel());
        }
    }

    /** An empty string is not evidence. Null is what the client reads as missing. */
    private static String blankToNull(String value) {
        return (value == null || value.isBlank()) ? null : value.trim();
    }

    private static String firstPresent(String supplied, String existing) {
        return supplied != null ? supplied : existing;
    }

    private static boolean hasText(String value) {
        return value != null && !value.isBlank();
    }
}
